### PR TITLE
WIP/PERF: block-wise ops for frame-with-series axis=1

### PR DESCRIPTION
**Pull request by jbrockmendel**
xref #32779 which does the same thing for frame-with-frame ops.  This one is turning out to be the most difficult of the cases to get right.

Posting this for comments, in large part because there are still two failing tests locally, and these may just be behaviors we want to change.

```
# test_scalar_na_logical_ops_corners
s = Series([2, 3, 4, 5, 6, 7, 8, 9, datetime(2005, 1, 1)])
s[::2] = np.nan
d = DataFrame({"A": s})

with pytest.raises(TypeError):
    d.__and__(s, axis="columns")   # <-- when done block-wise, this does not raise
```

The other one is harder to post an example of because it uses fixtures making it a PITA to copy/paste from the test, but the upshot is that `op(frame[mixed-float-dtypes], series[float64])` downcasts to the original dtypes ATM (since each column is effectively operating against a scalar) but when done blockwise they all come back float64.

### CODE REVIEW

File: pandas/core/array_algos/npcompat.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+"""
+Implementations of high-level numpy functions that are ExtensionArray-compatible.
+"""
+import numpy as np
+
+from pandas._typing import ArrayLike
+
+
+def tile(arr: ArrayLike, shape) -> ArrayLike:
+    raise NotImplementedError
```
Comment (TomAugspurger) [inline pandas/core/array_algos/npcompat.py:10]: Is this leftover debug stuff?

File: pandas/core/array_algos/npcompat.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    # TODO: if we are ndim==size==1 it shouldnt matter whether rowlike/columnlike?
+    if values.ndim == 1 and orient is not None:
+        # SUpport treating a 1-dimensional array as either a row or column
+        assert orient in ["rowlike", "columnlike"]
+        if orient == "rowlike":
+            values = values.reshape(1, -1)
+        else:
+            values = values.reshape(-1, 1)
+
+    btvalues = np.broadcast_to(values, shape)
```
Comment (TomAugspurger) [inline pandas/core/array_algos/npcompat.py:29]: Is the result `btvalues` being returned to a user? `broadcast_to` is typically (always?) a readonly view, and I don't think we want to be returning readonly results.

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -455,6 +455,12 @@ def ravel(self, *args, **kwargs):
         data = self._data.ravel(*args, **kwargs)
         return type(self)(data, dtype=self.dtype)
 
+    @property
+    def T(self):
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py:459]: Where is this used? In an only datetimelike context?

File: pandas/core/array_algos/npcompat.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    if isinstance(arr, np.ndarray):
+        result = btvalues
+    else:
+        result = type(arr)._from_factorized(btvalues, arr)
```
Comment (TomAugspurger) [inline pandas/core/array_algos/npcompat.py:33]: Was the result of that other discussion not to use `_from_factorized` outside of `pd.factorize`? Or was that just `_values_for_factorize`?

File: pandas/core/indexes/period.py [hunk 5]
```diff
@@ -350,6 +351,17 @@ def _int64index(self) -> Int64Index:
     # ------------------------------------------------------------------------
     # Index Methods
 
+    def __getitem__(self, key):
+        # PeriodArray.__getitem__ returns PeriodArray for 2D lookups,
+        #  so we need to issue deprecation warning and cast here
```
Comment (TomAugspurger) [inline pandas/core/indexes/period.py:356]: Are you planning to split this into it's own PR? Don't want to have it held up by review here.

File: pandas/core/array_algos/npcompat.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/array_algos/npcompat.py:10]: more or less, will remove

File: pandas/core/array_algos/npcompat.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/array_algos/npcompat.py:29]: the way its used in this PR, it is not returned to a user, but used as an intermediate object in the arithmetic op

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:459]: yes

File: pandas/core/indexes/period.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/indexes/period.py:356]: hadnt planned on it, but might as well.  this PR is back-burner until at least the frame-with-frame case is done

File: pandas/core/array_algos/npcompat.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/array_algos/npcompat.py:33]: That has more or less stalled.  joris made a good point that e.g. fletcher doesnt use _values_for_factorize, so ive come around to the opinion that we shouldn't have _values_for_factorize/_from_factorized at all, since they are each only used once in EA.factorize.

Then the issue becomes whether we can use _values_for_argsort, and if we have a constructor that can round-trip those.

**Comment by jbrockmendel** [comment]
mothballing to clear the queue
